com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/LockableStackFrameManager.java

```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package com.microsoft.java.debug.core.adapter;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.sun.jdi.IncompatibleThreadStateException;
+import com.sun.jdi.StackFrame;
+import com.sun.jdi.ThreadReference;
+
+public class LockableStackFrameManager implements ILockableStackFrameManager {
+    private Map<Long, StackFrame[]> threadStackFrameMap = Collections.synchronizedMap(new HashMap<>());
+    private Map<Long, ReentrantLock> locks = new HashMap<>();
+
+    @Override
+    public DisposableReentrantLock<StackFrame> getLockedStackFrame(ThreadReference thread, int depth) {
+        ReentrantLock lock = locks.computeIfAbsent(thread.uniqueID(), t -> new ReentrantLock());
+        lock.lock();
+        StackFrame sf = getStackFrame(thread, depth);
```

Comment:
No need to call another function.
```java
try {
    StackFrame sf = threadStackFrameMap.get(thread.uniqueID())[depth];
    return new LockedObject<>(sf, lock);
} catch (WhatEverException) {
    lock.unlock();
    throw;
}
```
The caller needs to handle the exception. Real scenario - when a queued evaluation request wants to acquire stack frame that is popped right after an HCR.